source/ota_mqtt.c

```diff
@@ -207,6 +215,117 @@ static uint32_t prvBuildStatusMessageFinish( char * pMsgBuffer,
                                              int32_t reason,
                                              int32_t subReason );
 
+/**
+ * @brief Build a string from a set of strings
+ *
+ * @param[in] pBuffer Buffer to place the output string in.
+ * @param[in] bufferSizeBytes Size of the buffer pointed to by pBuffer.
+ * @param[in] strings NULL-terminated array of string pointers.
+ * @return size_t Length of the output string, not including the terminator.
+ */
+static size_t stringBuilder( char * pBuffer,
+                             size_t bufferSizeBytes,
+                             const char * strings[] );
+
+/**
+ * @brief Build a string with the decimal representation of a uint32_t value.
+ *
+ * @param[in] pBuffer Buffer to place the output string in.
+ * @param[in] value The uint32_t value to convert.
+ * @return size_t Length of the output string, not including the terminator.
+ */
+static size_t stringBuilderUInt32Decimal( char * pBuffer,
+                                          uint32_t value );
+
+/**
+ * @brief Build a string with the hex representation of a uint32_t value.
+ *
+ * @param[in] pBuffer Buffer to place the output string in.
+ * @param[in] value The uint32_t value to convert.
+ * @return size_t Length of the output string, not including the terminator.
+ */
+static size_t stringBuilderUInt32Hex( char * pBuffer,
+                                      uint32_t value );
+
+static size_t stringBuilder( char * pBuffer,
+                             size_t bufferSizeBytes,
+                             const char * strings[] )
+{
+    size_t curLen = 0;
+    int i;
+
+    pBuffer[ 0 ] = '\0';
+
+    for( i = 0; strings[ i ] != NULL; i++ )
+    {
+        strncat( pBuffer, strings[ i ], bufferSizeBytes - curLen - 1 );
+        curLen += strlen( strings[ i ] );
+    }
+
+    return curLen;
+}
+
+static size_t stringBuilderUInt32Decimal( char * pBuffer,
+                                          uint32_t value )
+{
+    static const char decDigits[] =
+    {
+        '0', '1', '2', '3',
+        '4', '5', '6', '7',
+        '8', '9'
+    };
+    char workBuf[ U32_MAX_LEN ];
+    char * pCur = workBuf;
+    char * pDest = pBuffer;
+    size_t size = 0;
+
+    while( value )
+    {
+        *pCur++ = decDigits[ ( value % 10 ) ];
```

Comment:
What was the reasoning here to not just do *pCur++ = '0' + ( value % 10 );

Reply:
That's what the first version did, but it gives compiler warnings for `int` -> `char` conversions so I changed to this approach.